File: c_version/code/lz78.c
//
//  lz78.c
//
//  Created by Or Leibovich, Yochai Meir, and Itai Sharon, last updated on 11/Sep/22
//

#include "lz78.h"
#include <assert.h>

#define	CHECK_BIT(idx)	((lz->mem[(idx) >> 3] & (128 >> ((idx) & 7))) > 0)

////////////////////////////////////////////////////////////////////////////////
// LZ78
////////////////////////////////////////////////////////////////////////////////
struct s_LZ78
{
    unsigned char *mem;		// Bit array that keeps the inner nodes
    unsigned int mem_size;	// Memory size allocated for mem
    unsigned int max_depth;	// Maximum depth including leaves, specified by the user. 
    unsigned int leaf_count;	// Total number of leaves (paths) in the tree. Used for calculating log-loss
    unsigned int full_depth;	// Maximum depth in which all inner nodes are present
    char* name;
    size_t num_nodes_in_depth[MAX_DEPTH+1];	// Keeps the number of inner nodes in each depth
};

////////////////////////////////////////////////////////////////////////////////
void    LZ78WriteStats(LZ78 lz, FILE* fout) {
    fprintf(fout, "Name:                      %s\n", LZ78Name(lz));
    fprintf(fout, "Node array size:           %u\n", lz->mem_size);
    fprintf(fout, "Number of inner nodes:     %u\n", LZ78NumInnerNodes(lz));
    fprintf(fout, "Max complete depth:        %u\n", LZ78MaxCompleteDepth(lz));
    fprintf(fout, "Longest path (root->leaf): %u\n", LZ78LongestPathRootToLeaf(lz));
    fprintf(fout, "Number of inner node in each depth (%c of possible nodes):\n", '%');
    fprintf(fout, "Depth\tNNodes\tNFull\t%c of full\n", '%');

    fprintf(fout, "0\t1\t1\t100.0\n");
    for(int i=1; i<lz->max_depth; i++) {
       size_t full_n = ((size_t)4) << 2*(i-1);
       fprintf(fout, "%d\t%lu\t%lu\t%.1lf\n", i, lz->num_nodes_in_depth[i], full_n, 100.0*((double)lz->num_nodes_in_depth[i])/((double) full_n));
    }
    fprintf(fout, "\nNumber of leaves:\t%u\n", lz->leaf_count);
}

////////////////////////////////////////////////////////////////////////////////
// len_bases contains the number of bits required for each depth. At each depth, we need 4 nodes for each leaf
// in the previous depth. depth 1: 4, depth 2: 4+16, depth 3: 4+16+64, etc. Maximum depth is 17
size_t len_bases[MAX_DEPTH+1] = {}; // {0, 4, 20, 84, 340, 1364, 5460, 21844, 87380, 349524, 1398100, 5592404, 22369620, 89478484, 357913940, 1431655764, 5726623060, 22906492244};

// Returns the memory size required for keeping all inner nodes up to depth max_depth,
// where max_depth may contain only leaves
size_t calc_mem_size(unsigned int max_depth)
{
    return (len_bases[max_depth-1] / 8) + 1;
}

////////////////////////////////////////////////////////////////////////////////
void AddNode(LZ78 lz, size_t node_index, unsigned int depth)
{
    lz->mem[node_index >> 3] |= (128 >> (node_index & 7));
    lz->num_nodes_in_depth[depth]++;

    // One leaf became an inner node, 4 new leaves were added, 3 new leaves added in total
    lz->leaf_count += 3;
}

////////////////////////////////////////////////////////////////////////////////
LZ78 LZ78Create(const char *name, unsigned int depth)
{
    if((depth == 0) || (depth > MAX_DEPTH)) {
        fprintf(stderr, "%s, %d: requested depth (%u) is illegal, must be between 1 to %d. Aborting\n\n", __FILE__, __LINE__, depth, MAX_DEPTH);
        exit(-1);
    }

    LZ78 lz = (LZ78)calloc(1, sizeof(struct s_LZ78));
    if (!lz)
    {
        fprintf(stderr, "%s, %d: memory allocation of %lu bytes failed\n\n", __FILE__, __LINE__, sizeof(struct s_LZ78));
        exit(-1);
    }

    lz->name = strdup(name);
    if(!lz->name) {
        fprintf(stderr, "%s, %d: str duplication failed\n\n", __FILE__, __LINE__);
        exit(-1);
    }

    // len_bases is the starting index for each depth i AND the number of indices
    // required for depth i-1     
    for(int i=1; i<=MAX_DEPTH; i++)
        len_bases[i] = len_bases[i-1] + (4 << 2*(i-1));

    lz->mem_size = calc_mem_size(depth);
    lz->max_depth = depth;
    lz->mem = (unsigned char *)calloc(lz->mem_size, sizeof(unsigned char));
    if (!lz->mem)
    {
        fprintf(stderr, "%s, %d: memory allocation of %lu bytes failed\n\n", __FILE__, __LINE__, lz->mem_size*sizeof(unsigned char));
        exit(-1);
    }

    lz->num_nodes_in_depth[0] = 1;
    lz->leaf_count = 4;
    lz->full_depth = 0;

    return lz;
}

////////////////////////////////////////////////////////////////////////////////
void LZ78Destroy(LZ78 lz) {
    free(lz->mem);
    free(lz->name);
    free(lz);
}

////////////////////////////////////////////////////////////////////////////////
// For a sequence s=s1s2...sn, the index in the bit memory can be calculated 
// using m(s1..si) = (4^0+..+4^(i-1))-1 + 4*m(s1..si-1) 
// There is some bug here
size_t LZ78Build(LZ78 lz, const char *file_path)
{
    File file = FileCreate(file_path);

    char *buf = NULL;
    size_t buf_size = 0;
    size_t n;

    unsigned int curr_depth = 1; // len is at least 1
    size_t curr_sequence = 0;  // sequence value.

    while ((n = FileGetline(&buf, &buf_size, file)) != 0) // EOF)
    {
        if ((*buf == '>') || (*buf == '\n'))
        {
            curr_depth = 1;
            curr_sequence = 0;
            continue;
        }
        char *p = buf, *pe = buf + n;

        // buf may or may not end with \n
	if(*(pe-1) == '\n')
		pe--;

        for (; p < pe; p++)
        {
            // toupper
            if (*p >= 'a')
                (*p) -= 32;

            // This will start a new path
            if (*p == 'N')
            {
                curr_depth = 1;
                curr_sequence = 0;
                continue;
            }

            // For (*p >> 1) this is what we get:
            // A = 0b100000
            // C = 0b100001
            // G = 0b100011
            // T = 0b101010
            // Order will be ACTG
            int i = (*p >> 1) & 3;
            curr_sequence |= i;

            // As far as I understand this can only happen if lz->max_depth == 1
            if(curr_depth > lz->max_depth-1) {
                curr_depth = 1;
                curr_sequence = 0;
                continue;
            }

            size_t current_index = len_bases[curr_depth - 1] + curr_sequence;
            assert((current_index >> 3) < lz->mem_size);
            if (!CHECK_BIT(current_index))
            {
                AddNode(lz, current_index, curr_depth);
                curr_depth = 1;
                curr_sequence = 0;
                continue;
            }

            if(curr_depth == lz->max_depth-1) {
                curr_depth = 1;
                curr_sequence = 0;
            }
            else
            {
                curr_sequence <<= 2;
                curr_depth++;
            }
        }
    }

    // Check what is the maximum level with all nodes
    lz->full_depth = 0;
    while((lz->full_depth+1 < lz->max_depth) && (lz->num_nodes_in_depth[lz->full_depth+1] == (4 << 2*(lz->full_depth))))
        lz->full_depth++;
    
    free(buf);
    FileDestroy(file);
    return 0;
}

////////////////////////////////////////////////////////////////////////////////
double LZ78AverageLogScore(LZ78 lz, File file)
{
    char *buf = NULL;
    size_t buf_size = 0;
    size_t n;

    unsigned int nchars = 0;
    unsigned int actual_nchars = 0;
    unsigned int leaf_count = 0;

    unsigned int curr_depth = 1;
    size_t curr_sequence = 0;

    FileRollBack(file);

    while ((n = FileGetline(&buf, &buf_size, file)) != 0) // EOF)
    {
        if ((*buf == '>') || (*buf == '\n'))
        {
            curr_depth = 1;
            curr_sequence = 0;
            continue;
        }
        char *p = buf, *pe = buf + n;

        // buf may or may not end with \n
	if(*(pe-1) == '\n')
		pe--;

        for (; p < pe; p++)
        {
            if (*p >= 'a')
                (*p) -= 32;
            if (*p == 'N')
            {
                curr_depth = 1;
                curr_sequence = 0;
                continue;
            }
            // For (*p >> 1) this is what we get:
            // A = 0b100000
            // C = 0b100001
            // G = 0b100011
            // T = 0b101010
            // Order will be ACTG
            int i = (*p >> 1) & 3;
            nchars++;
            curr_sequence |= i;
            size_t current_index = len_bases[curr_depth - 1] + curr_sequence;

            // The first part is an optimization: no need to check if the node 
            // exists for depths with all inner nodes present 
            if ((curr_depth <= lz->full_depth) || ((curr_depth < lz->max_depth) && CHECK_BIT(current_index)))
            {
                curr_sequence <<= 2;
                curr_depth++;
            }
            else
            {
                leaf_count++;
                curr_depth = 1;
                curr_sequence = 0;
                actual_nchars = nchars;
                continue;
            }
        }
    }
    free(buf);
    return (log(lz->leaf_count) / log(2) * leaf_count) / actual_nchars; // log_2
}

////////////////////////////////////////////////////////////////////////////////
const char *LZ78Name(LZ78 lz)
{
    return lz->name;
}

////////////////////////////////////////////////////////////////////////////////
unsigned int LZ78NumInnerNodes(LZ78 lz)
{
    size_t n = 0;
    for(int i=0; i<lz->max_depth; i++)
        n += lz->num_nodes_in_depth[i];

    return n;
}
////////////////////////////////////////////////////////////////////////////////
unsigned int LZ78MaxCompleteDepth(LZ78 lz)
{
    return lz->full_depth;
}

////////////////////////////////////////////////////////////////////////////////
unsigned int LZ78LongestPathRootToLeaf(LZ78 lz)
{
    unsigned int i;
    for(i=0; i<lz->max_depth; i++)
        if(lz->num_nodes_in_depth[i] == 0)
            break;
    return i;
}

File: code/src/lz78.rs
//
//  lz78.rs
//
//  Created by Or Leibovich, Yochai Meir, and Itai Sharon, last updated on 11/Sep/22
//

use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;


pub struct LenBases {
    len_bases: Vec<usize>,
}

impl LenBases  {
    pub fn new(max_depth: usize) -> Self {
        let mut len_bases = vec![0_usize; max_depth + 1]; // +1
        // // len_bases is the starting index for each depth i AND the number of indices
        // // required for depth i-1
        // {0, 4, 20, 84, 340, 1364, 5460, 21844, 87380, 349524, 1398100, 5592404, 22369620, 89478484, 357913940, 1431655764, 5726623060, 22906492244};
        for i in 1..=max_depth {
            len_bases[i] = len_bases[i-1] + (4 << (2*(i-1)));
        }

        LenBases {
            len_bases
        }
    }

    pub fn bases(&self) -> &[usize] { &self.len_bases }
}

// Returns the memory size required for keeping all inner nodes up to depth max_depth,
// where max_depth may contain only leaves
fn calc_mem_size(len_bases: &LenBases, max_depth: usize) -> usize {
    (len_bases.bases()[max_depth-1] / 8) + 1
}


////////////////////////////////////////////////////////////////////////////////
// LZ78
////////////////////////////////////////////////////////////////////////////////
pub struct LZ78<'a> {
    mem: Vec<u8>, // Bit array that keeps the inner nodes
    mem_size: usize, // Memory size allocated for mem
    max_depth: usize, // Maximum depth including leaves, specified by the user.
    leaf_count: usize, // Total number of leaves (paths) in the tree. Used for calculating log-loss
    full_depth: usize, // Maximum depth in which all inner nodes are present
    name: String,
    num_nodes_in_depth: Vec<usize>, // Keeps the number of inner nodes in each depth
    len_bases: &'a LenBases,
    buffer_size: usize,
}

impl<'a> LZ78<'a> {
    fn check_bit(&self, idx: usize) -> bool {
        (self.mem[idx >> 3] & (128_u8 >> (idx as u8 & 7))) > 0
    }

    pub fn get_name(&self) -> &str { self.name.as_str() }

    pub fn new(name: &str, max_depth: usize, len_bases: &'a LenBases, file_path: &Path, buffer_size: usize) -> Self {
        let mem_size = calc_mem_size(len_bases, max_depth);

        let mut lz = LZ78 {
            mem: vec![0_u8; mem_size],
            max_depth,
            mem_size,
            name: name.to_string(),
            leaf_count: 4,
            full_depth: 0,
            num_nodes_in_depth: vec![0; max_depth+1],
            len_bases,
            buffer_size,
        };
        lz.num_nodes_in_depth[0] = 1;
        lz.build(file_path);
        lz
    }


    // For a sequence s=s1s2...sn, the index in the bit memory can be calculated
    // using m(s1..si) = (4^0+..+4^(i-1))-1 + 4*m(s1..si-1)
    // There is some bug here
    fn build(&mut self, file_path: &Path) {
        let mut curr_depth = 1; // len is at least 1
        let mut curr_sequence: usize = 0_usize;  // sequence value.

        let mut file_reader = BufReader::new(File::open(file_path)
            .unwrap_or_else(|_| panic!("E: failed to read {}", file_path.display())));

        let mut bytes_buffer = vec![0_u8; self.buffer_size];
        let mut desc_line = false;

        loop {
            match file_reader.read(&mut bytes_buffer) {
                Ok(0) => break,
                Ok(n) => {
                    for p in &bytes_buffer[0..n] {
                        if *p == b'>' {
                            desc_line = true;
                            curr_depth = 1;
                            curr_sequence = 0;
                        } else if desc_line {
                            if *p == b'\n' {
                                desc_line = false;
                            }
                        } else if *p != b'\n' {
                            let p = if *p >= b'a' {
                                p - 32
                            } else {
                                *p
                            };

                            if p == b'N' {
                                curr_depth = 1;
                                curr_sequence = 0;
                                continue;
                            }

                            // For (p >> 1) this is what we get:
                            // A = 0b100000
                            // C = 0b100001
                            // G = 0b100011
                            // T = 0b101010
                            // Order will be ACTG
                            curr_sequence |= ((p >> 1) & 3) as usize;

                            // As far as I understand this can only happen if self.max_depth == 1
                            if curr_depth > self.max_depth-1 {
                                curr_depth = 1;
                                curr_sequence = 0;
                                continue;
                            }

                            //
                            let current_index = self.len_bases.bases()[curr_depth - 1] + curr_sequence;
                            assert!((current_index >> 3) < self.mem_size);
                            if !self.check_bit(current_index) {
                                self.add_node(current_index, curr_depth);
                                curr_depth = 1;
                                curr_sequence = 0;
                                continue;
                            }

                            if curr_depth == self.max_depth - 1 {
                                curr_depth = 1;
                                curr_sequence = 0;
                            } else {
                                curr_sequence <<= 2;
                                curr_depth += 1;
                            }
                        }
                    }
                },
                Err(e) => panic!("E: failed reading file {}, got {:?}", file_path.display(), e),
            };
        }

        // Check what is the maximum level with all nodes
        self.full_depth = 0;
        while self.full_depth + 1 < self.max_depth && self.num_nodes_in_depth[self.full_depth + 1] == (4 << (2 * self.full_depth)) {
            self.full_depth += 1;
        }
    }

    fn add_node(&mut self, node_index: usize, depth: usize) {
        self.mem[node_index >> 3] |= 128 >> (node_index & 7);
        self.num_nodes_in_depth[depth] += 1;

        // One leaf became an inner node, 4 new leaves were added, 3 new leaves added in total
        self.leaf_count += 3;
    }

    pub fn average_log_score(&self, file_path: &Path) -> f64 {
        let mut nchars= 0;
        let mut actual_nchars: usize = 0;
        let mut leaf_count: usize = 0;

        let mut curr_depth: usize = 1;
        let mut curr_sequence: usize = 0;

        let mut file_reader = BufReader::new(File::open(file_path)
            .unwrap_or_else(|_| panic!("E: failed to read {}", file_path.display())));

        let mut bytes_buffer = vec![0_u8; self.buffer_size];
        let mut desc_line = false;

        loop {
            match file_reader.read(&mut bytes_buffer) {
                Ok(0) => break,
                Ok(n) => {
                    for p in &bytes_buffer[0..n] {
                        if *p == b'>' {
                            desc_line = true;
                            curr_depth = 1;
                            curr_sequence = 0;
                        } else if desc_line {
                            if *p == b'\n' {
                                desc_line = false;
                            }
                        } else if *p != b'\n' {
                            // To upper
                            let p = if *p >= b'a' {
                                p - 32
                            } else {
                                *p
                            };
                            if p == b'N' {
                                curr_depth = 1;
                                curr_sequence = 0;
                                continue;
                            }

                            // For (p >> 1) this is what we get:
                            // A = 0b100000
                            // C = 0b100001
                            // G = 0b100011
                            // T = 0b101010
                            // Order will be ACTG
                            let i = (p as usize >> 1) & 3;
                            nchars += 1;
                            curr_sequence |= i;
                            let current_index = self.len_bases.bases()[curr_depth - 1] + curr_sequence;

                            // The first part is an optimization: no need to check if the node
                            // exists for depths with all inner nodes present
                            if curr_depth <= self.full_depth || (curr_depth < self.max_depth && self.check_bit(current_index)) {
                                curr_sequence <<= 2;
                                curr_depth += 1;
                            } else {
                                leaf_count += 1;
                                curr_depth = 1;
                                curr_sequence = 0;
                                actual_nchars = nchars;
                                continue;
                            }
                        }
                    }
                },
                Err(e) => panic!("E: failed reading file {}, got {:?}", file_path.display(), e),
            };
        }

        (self.leaf_count as f64).log2() * leaf_count as f64 / actual_nchars as f64
    }

    fn num_inner_nodes(&self) -> usize {
        (0..self.max_depth)
            .map(|i| self.num_nodes_in_depth[i])
            .sum()
    }

    fn max_complete_depth(&self) -> usize
    {
        self.full_depth
    }

    fn longest_path_root_to_leaf(&self) -> usize {
        let node = self.num_nodes_in_depth.iter()
            .enumerate()
            .find(|(_, &num_nodes_in_depth)| num_nodes_in_depth == 0);

        if let Some((i, _)) = node {
            i
        } else {
            self.max_depth
        }
    }
}

impl<'a> std::fmt::Display for LZ78<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Name:                      {}\n\
                   Node array size:           {}\n\
                   Number of inner nodes:     {}\n\
                   Max complete depth:        {}\n\
                   Longest path (root->leaf): {}\n\
                   Number of inner node in each depth (% of possible nodes):\n\
                   Depth\tNNodes\tNFull\t% of full\n\
                   0\t1\t1\t100.0\n", self.name,
               self.mem_size,
               self.num_inner_nodes(),
               self.max_complete_depth(),
               self.longest_path_root_to_leaf())?;

        for i in 1..self.max_depth {
            let full_n = 4_usize << (2*(i-1));
            writeln!(f, "{}\t{}\t{}\t{:.1}", i,
                                             self.num_nodes_in_depth[i],
                                             full_n,
                                             100.0*(self.num_nodes_in_depth[i] as f64)/(full_n as f64))?;
        }
        writeln!(f, "\nNumber of leaves:\t{}", self.leaf_count)
    }
}

